### DISCUSSION TITLE
Advice on rendering JSON in a controller when multiple things need to be rendered

### DISCUSSION THREAD

**Discussion by Sheikh-TDH** [okuramasafumi/alba#254] [Q&A] [answered]
Currently I have controllers rendering multiple things from an action for example

`render json: { items: items, total_entries: @foo.count, incomplete_results: false }, status: 200`

I am currently utilizing the serializable_hash method and that seems to work fine, but it was pointed out to me that JSON to a hash and then back to json is taxing so I'm switching to serialize instead. The problem is since serialize produces raw json the above bit of code no longer works.

I have a current implementation working for this 
```
    json_template = '{"items": %{items}, "total_entries": %{total_entries}, "incomplete_results": %{incomplete_results}}'
    raw_json = json_template % {items: FooResource.new(paginated_foo).serialize, total_entries: @foo.count, incomplete_results: false}
    render json: raw_json, status: 200
```
But I wanted advice if there is a better way on going about doing this

And also as a side question what is the most performant method between serialize and serializable_hash?

**Comment by okuramasafumi** [comment]
@Sheikh-TDH Does `layout` feature work for you?
https://github.com/okuramasafumi/alba#layout

> And also as a side question what is the most performant method between serialize and serializable_hash?
`serializable_hash` is faster because `serialize` method uses `serializable_hash` internally.

**Reply by Sheikh-TDH** [reply]
No the layout wont work after testing it out. Previously with ASM we were doing something like:

```items = ActiveModel::Serializer::CollectionSerializer.new(foo, serializer: FooSerializer)``` which would create a serializer object but when
doing ```render json: { items: items, total_entries: @foo.count, incomplete_results: false }, status: 200```

th json string that was produced had everything in the hash above.

**Reply by okuramasafumi** [reply]
OK, I'll investigate.
Could you provide a valid test case? It doesn't have to be `MiniTest::Case`, just input and output pair is enough.

**Reply by Sheikh-TDH** [reply]
For ASM?

given `render json: { items: items, total_entries: @foo.count, incomplete_results: false }, status: 200`

items = 
```#<ActiveModel::Serializer::CollectionSerializer:0x00007fbbaebd9c98
 @object=
  [#<Foo:0x00007fbbaebd7cb8
    id: 109,
    foo_date: Sun, 30 Oct 2022 15:00:00.000000000 CDT -05:00,
    created_by: nil,
    status: "processing",
    created_at: Sat, 15 Oct 2022 11:07:58.312297000 CDT -05:00,
    updated_at: Sat, 15 Oct 2022 11:07:58.312297000 CDT -05:00],
 @options={:serializer=>FooSerializer},
 @root=nil,
 @serializers=
  [#<FooSerializer:0x00007fbbaeb27a98
    @instance_options={},
    @object=
     #<Foo:0x00007fbbaebd7cb8
      id: 109,
      foo_date: Sun, 30 Oct 2022 15:00:00.000000000 CDT -05:00,
      created_by: nil,
      status: "processing",
      created_at: Sat, 15 Oct 2022 11:07:58.312297000 CDT -05:00,
      updated_at: Sat, 15 Oct 2022 11:07:58.312297000 CDT -05:00>,
    @root=nil,
    @scope=nil>]>```

@foo.count = 1

when `render json: { items: items, total_entries: @foo.count, incomplete_results: false }, status: 200` is executed this you get

```"{\"items\":[{\"id\":109,\"status\":\"processing\",\"foo_date\":\"2022-10-30T15:00:00.000-05:00\",\"foo_number\":\"102022374301\",\"created_at\":1665850078312.297,\"invoices\":[]}],\"total_entries\":1,\"incomplete_results\":false}"```

**Reply by Sheikh-TDH** [reply]
I have tried not adding the serialize method (or the serializable_hash method) but its spotty. In some areas of the app it works others it doesnt. For example if I do for this example, it works (it does give the deprecation warning `[DEPRECATION] `to_hash` is deprecated, use `serializable_hash` instead.`)

```
 items = FooSerializer.new(paginated_foo)
    render json: { items: items, total_entries: @foo.count, incomplete_results: false }, status: 200
```

but for this one it will fail:

`render json: BarSerializer.new(data), status: 200, layout: nil`

unless you do

`render json: {items: BarSerializer.new(data)}, status: 200, layout: nil`

I could go the latter route but I dont know if thats recommended

But I will say that not calling serializable or serializable hash and letting rails render the json is BLAZING fast look at the comparision below:

BarSerializer.new(data):

```
#<Benchmark::Tms:0x00007f06b67aed10
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=8.199999956559623e-05,
 @stime=1.3000000000040757e-05,
 @total=8.900000000000574e-05,
 @utime=7.599999999996498e-05>
```

BarSerializer.new(data).serializable_hash:

```#<Benchmark::Tms:0x00007f06b6854850
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=0.0035996999999952095,
 @stime=0.0005069999999998132,
 @total=0.0034220000000013684,
 @utime=0.0029150000000015552>```

**Reply by okuramasafumi** [reply, marked as answer]
Now I think your current solution using `serializable_hash` would be good enough.

> it was pointed out to me that JSON to a hash and then back to json is taxing

`serializable_hash` just returns a Hash so there is not cost of converting objects into a JSON. You can build a Hash with multiple Hashes some of which are built by `serializable_hash` and then convert it to JSON only once.

**Reply by Sheikh-TDH** [reply]
Cool! Thanks for clarifying
